compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantTypeChecker.java

```diff
@@ -549,7 +550,24 @@ private void reportIncompatibleMappingConstructorError(BLangRecordLiteral mappin
         if (expType.tag != TypeTags.UNION) {
             dlog.error(mappingConstructorExpr.pos,
                     DiagnosticErrorCode.MAPPING_CONSTRUCTOR_COMPATIBLE_TYPE_NOT_FOUND, expType);
+            return;
+        }
+
+        BUnionType unionType = (BUnionType) expType;
+        BType[] memberTypes = types.getAllTypes(unionType, true).toArray(new BType[0]);
+
+        // By this point, we know there aren't any types to which we can assign the mapping constructor. If this is
+        // case where there is at least one type with which we can use mapping constructors, but this particular
+        // mapping constructor is incompatible, we give an incompatible mapping constructor error.
+        for (BType bType : memberTypes) {
```

Comment:
Why are we duplicating the logic in `reportIncompatibleMappingConstructorError()` in `TypeChecker.java` here?

Can we use the same in the ConstantTypeChecker? 
The only difference is special casing the `T?`. Isn't it useful for constants as well?

Sample:
```bal
type T record {|
    string s;
    int i?;
|};

T? f1 = {s: "foo", invalid: 123}; // [1]
const T? f1 = {s: "foo", invalid: 123}; // [2]
```
For [2] also can give the same diagnostics as [1] which is more user-friendly.

Reply:
Add a comment as we discussed offline https://github.com/ballerina-platform/ballerina-lang/pull/40799/files#diff-b4cd70391128b02ba62b9a69264ea2010300b8dc1abe6e1f80a4c24bd70491e9R543